### PR TITLE
Feature/podspec v2

### PR DESCRIPTION
**Pull request by ycliuhw**
## Description of change

**this a draft, need further discussion;**

- deprecate current podspec to `legacy` version;
- introduce podspec `v2`;

## QA steps

- bootstrap;
- deploy CaaS workloads;

```bash
$ juju add-model t1 microk8s
$ juju deploy cs:~juju/mariadb-k8s-0 --debug
$ juju deploy cs:~juju/mediawiki-k8s-3 --debug
$ juju relate mediawiki-k8s:db mariadb-k8s:server
$ juju expose mediawiki-k8s
```

## Documentation changes

Yes

## Bug reference

None

### CODE REVIEW

File: caas/broker.go [hunk 1]
```diff
@@ -142,8 +143,8 @@ type Broker interface {
 	// DeleteOperator deletes the specified operator.
 	DeleteOperator(appName string) error
 
-	// EnsureCustomResourceDefinition creates or updates a custom resource definition resource.
-	EnsureCustomResourceDefinition(appName string, podSpec *PodSpec) error
+	// // EnsureCustomResourceDefinition creates or updates a custom resource definition resource.
```
Comment (wallyworld) [inline caas/broker.go]: Fix me

File: caas/kubernetes/provider/k8s.go [hunk 2]
```diff
@@ -1231,16 +1233,37 @@ func (k *kubernetesClient) EnsureService(
...
 	}
+
+	// ensure custom resource definitions first,
+	crds := svcSpec.CustomResourceDefinitions
+	if len(crds) > 0 {
+		err = k.ensureCustomResourceDefinitions(appName, crds)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		// TODO: add cleanups for crds !!!!!!!!!!!!!!!!!!
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: not enough !!!!!!, needs a few more

File: caas/kubernetes/provider/k8s.go [hunk 3]
```diff
@@ -2570,59 +2586,65 @@ func operatorConfigMap(appName, operatorName string, config *caas.OperatorConfig
 	}
 }
 
-type unitSpec struct {
+// TODO: move this to ./specs??????????
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: I think so. prepareSvcSpec() should probably move as well

File: caas/kubernetes/provider/k8s_test.go [hunk 4]
```diff
@@ -49,56 +51,77 @@ type K8sSuite struct {
...
+
+	podSpec.ProviderPod = &k8sspecs.K8sPodSpec{
+		ActiveDeadlineSeconds: int64Ptr(10),
+		KubernetesResources: &k8sspecs.KubernetesResources{
+			ServiceAccount: &k8sspecs.ServiceAccountSpec{
+				Name:                         "serviceAccount",
+				AutomountServiceAccountToken: boolPtr(true),
+			},
+		},
+		Hostname:                      "host",
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s_test.go]: We are removing Hostname, Subdomain, even for v1

File: caas/kubernetes/provider/specs/legacy.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	core "k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	k8syaml "k8s.io/apimachinery/pkg/util/yaml"
+
+	// "github.com/juju/juju/caas"
+	"github.com/juju/juju/caas/specs"
+)
+
+type podSpecLegacy struct {
+	CaaSSpec specs.PodSpecLegacy
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/legacy.go]: caasSpec

do these need to be exported?

File: caas/kubernetes/provider/specs/legacy.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+	if err := p.K8sSpec.Validate(); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (p podSpecLegacy) ToLatest() *specs.PodSpec {
+	pSpec := &specs.PodSpec{}
+	pSpec.Version = specs.CurrentVersion
+	pSpec.OmitServiceFrontend = p.CaaSSpec.OmitServiceFrontend
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/legacy.go]: OmitServiceFrontend becomes part of the deployment service type defined in charm metadata.yaml

File: caas/kubernetes/provider/specs/legacy_test.go [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+		}
+		return pSpecs
+	}
+
+	spec, err := k8sspecs.ParsePodSpec(specStrBase)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(spec, jc.DeepEquals, getExpectedPodSpecBase())
+}
+
+func float64Ptr(f float64) *float64 {
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/legacy_test.go]: there are defined in a couple of places, can we consolidate?

File: caas/kubernetes/provider/specs/v2.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+			return errors.Trace(err)
+		}
+	}
+	return nil
+}
+
+// KubernetesResources is the k8s related resources.
+type KubernetesResources struct {
+	// TODO: move secret, cm etc here for V2 ??????????
+	ServiceAccount            *ServiceAccountSpec                                          `json:"serviceAccount,omitempty"`
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/v2.go]: service account is modelled generically

File: caas/kubernetes/provider/specs/legacy.go (hunk 6 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/specs/legacy.go]: left a TOD to remove `OmitServiceFrontend` from v3 when we deprecate the legacy version.

File: caas/kubernetes/provider/specs/legacy_test.go (hunk 7 above)
Comment (ycliuhw) [reply caas/kubernetes/provider/specs/legacy_test.go]: added a TODO to move these to testing package.

File: caas/kubernetes/provider/k8s.go [hunk 9]
```diff
@@ -1703,8 +1729,8 @@ func (k *kubernetesClient) deleteDeployment(name string) error {
 }
 
 func (k *kubernetesClient) configureStatefulSet(
-	appName, deploymentName, randPrefix string, annotations k8sannotations.Annotation, unitSpec *unitSpec,
-	containers []caas.ContainerSpec, replicas *int32, filesystems []storage.KubernetesFilesystemParams,
+	appName, deploymentName, randPrefix string, annotations k8sannotations.Annotation, unitSpec *workloadSpec,
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: Change the param name unitSpec to workloadSpec ?

File: caas/kubernetes/provider/k8s.go [hunk 10]
```diff
@@ -1644,8 +1670,8 @@ func podAnnotations(annotations k8sannotations.Annotation) k8sannotations.Annota
 func (k *kubernetesClient) configureDeployment(
 	appName, deploymentName string,
 	annotations k8sannotations.Annotation,
-	unitSpec *unitSpec,
-	containers []caas.ContainerSpec,
+	unitSpec *workloadSpec,
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: Change the param name unitSpec to workloadSpec ?

File: caas/kubernetes/provider/k8s.go [hunk 11]
```diff
@@ -2619,59 +2645,89 @@ func operatorConfigMap(appName, operatorName string, config *caas.OperatorConfig
...
+			cs.Containers = append(cs.Containers, c)
+		}
+	}
+
 	// Fill out the easy bits using a template.
 	var buf bytes.Buffer
-	if err := defaultPodTemplate.Execute(&buf, podSpec); err != nil {
-		return nil, errors.Trace(err)
+	if err := defaultPodTemplate.Execute(&buf, cs); err != nil {
+		logger.Errorf("unable to execute template for containers: %+v, err: %+v", cs, err)
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: Unless we really need to log, better to errors.Annotate() the error with this message

File: caas/kubernetes/provider/k8s.go [hunk 12]
```diff
@@ -2619,59 +2645,89 @@ func operatorConfigMap(appName, operatorName string, config *caas.OperatorConfig
...
+		return errors.Trace(err)
 	}
-	unitSpecString := buf.String()
 
-	var unitSpec unitSpec
+	unitSpecString := buf.String()
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: s/unitSpec/workloadSpec?
(here and below....)

File: caas/kubernetes/provider/k8s_test.go [hunk 13]
```diff
@@ -49,35 +51,52 @@ type K8sSuite struct {
...
+
+func int64Ptr(i int64) *int64 {
+	return &i
+}
+
+func boolPtr(b bool) *bool {
+	return &b
+}
+
+func (s *K8sSuite) TestMakeSvcSpecNoConfigConfig(c *gc.C) {
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s_test.go]: s/ConfigConfig/Config ?

File: caas/kubernetes/provider/k8s_test.go [hunk 14]
```diff
@@ -49,35 +51,52 @@ type K8sSuite struct {
...
+	}
+
+	podSpec.ProviderPod = &k8sspecs.K8sPodSpec{
+		Pod: &k8sspecs.PodSpec{
+			RestartPolicy:                 core.RestartPolicyOnFailure,
 			ActiveDeadlineSeconds:         int64Ptr(10),
-			ServiceAccountName:            "serviceAccount",
-			Hostname:                      "host",
-			Subdomain:                     "sub",
-			DNSPolicy:                     core.DNSClusterFirst,
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s_test.go]: Still need DNSPolicy, Priority, PriorityClassName as k8s specific attributes
I see DNSPolicy and Priority below, but not PriorityClassName?

File: caas/kubernetes/provider/k8s.go [hunk 15]
```diff
@@ -2619,59 +2645,89 @@ func operatorConfigMap(appName, operatorName string, config *caas.OperatorConfig
...
-		unitSpec.Pod.ActiveDeadlineSeconds = spec.ActiveDeadlineSeconds
-		unitSpec.Pod.ServiceAccountName = spec.ServiceAccountName
-		unitSpec.Pod.AutomountServiceAccountToken = spec.AutomountServiceAccountToken
-		unitSpec.Pod.TerminationGracePeriodSeconds = spec.TerminationGracePeriodSeconds
-		unitSpec.Pod.Hostname = spec.Hostname
-		unitSpec.Pod.Subdomain = spec.Subdomain
-		unitSpec.Pod.DNSConfig = spec.DNSConfig
-		unitSpec.Pod.DNSPolicy = spec.DNSPolicy
-		unitSpec.Pod.Priority = spec.Priority
-		unitSpec.Pod.PriorityClassName = spec.PriorityClassName
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s.go]: Still want Priority, PriorityClassName

File: caas/kubernetes/provider/k8s_test.go [hunk 16]
```diff
@@ -1099,7 +1116,8 @@ func (s *K8sBrokerSuite) TestEnsureServiceNoStorage(c *gc.C) {
 	defer ctrl.Finish()
 
 	numUnits := int32(2)
-	unitSpec, err := provider.MakeUnitSpec("app-name", "app-name", getBasicPodspec())
+	basicPodSpec := getBasicPodspec()
+	unitSpec, err := provider.PrepareWorkloadSpec("app-name", "app-name", basicPodSpec)
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s_test.go]: workloadSpec?
(here and elsewhere?

File: caas/kubernetes/provider/k8s_test.go [hunk 17]
```diff
@@ -388,8 +405,8 @@ func (s *K8sBrokerSuite) secretArg(c *gc.C, annotations map[string]string) *core
 	return secret
 }
 
-func (s *K8sSuite) TestMakeUnitSpecConfigPairs(c *gc.C) {
-	spec, err := provider.MakeUnitSpec("app-name", "app-name", getBasicPodspec())
+func (s *K8sSuite) TestMakeSvcSpecConfigPairs(c *gc.C) {
```
Comment (wallyworld) [inline caas/kubernetes/provider/k8s_test.go]: s/Svc/Workload?

File: caas/kubernetes/provider/specs/legacy.go [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+	pSpec.ProviderPod = &K8sPodSpec{
+		KubernetesResources: &KubernetesResources{
+			CustomResourceDefinitions: p.k8sSpec.CustomResourceDefinitions,
+		},
+		Pod: &PodSpec{
+			RestartPolicy:                 p.k8sSpec.RestartPolicy,
+			ActiveDeadlineSeconds:         p.k8sSpec.ActiveDeadlineSeconds,
+			TerminationGracePeriodSeconds: p.k8sSpec.TerminationGracePeriodSeconds,
+			SecurityContext:               p.k8sSpec.SecurityContext,
+			Priority:                      p.k8sSpec.Priority,
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/legacy.go]: Also PriorityClassName

File: caas/kubernetes/provider/specs/legacy_test.go [hunk 19]
```diff
@@ -0,0 +1,431 @@
...
+omitServiceFrontend: true
+activeDeadlineSeconds: 10
+restartPolicy: OnFailure
+terminationGracePeriodSeconds: 20
+automountServiceAccountToken: true
+serviceAccountName: serviceAccountFoo
+securityContext:
+  runAsNonRoot: true
+  supplementalGroups: [1,2]
+hostname: host
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/legacy_test.go]: Remove hostname, subdomain, dnsConfig etc

File: caas/kubernetes/provider/specs/types.go [hunk 20]
```diff
@@ -0,0 +1,183 @@
+// Copyright 2018 Canonical Ltd.
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/types.go]: 2019

File: caas/kubernetes/provider/specs/types.go [hunk 21]
```diff
@@ -0,0 +1,183 @@
...
+	return result
+}
+
+// K8sContainerSpec is a subset of v1.Container which defines
+// attributes we expose for charms to set.
+type K8sContainerSpec struct {
+	LivenessProbe   *core.Probe           `json:"livenessProbe,omitempty"`
+	ReadinessProbe  *core.Probe           `json:"readinessProbe,omitempty"`
+	SecurityContext *core.SecurityContext `json:"securityContext,omitempty"`
+	ImagePullPolicy core.PullPolicy       `json:"imagePullPolicy,omitempty"`
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/types.go]: I think ImagePullPolicy is ok in the generic bit

File: caas/kubernetes/provider/specs/types.go [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+}
+
+// PodSpec is a subset of v1.PodSpec which defines
+// attributes we expose for charms to set.
+type PodSpec struct {
+	RestartPolicy                 core.RestartPolicy       `json:"restartPolicy,omitempty"`
+	ActiveDeadlineSeconds         *int64                   `json:"activeDeadlineSeconds,omitempty"`
+	TerminationGracePeriodSeconds *int64                   `json:"terminationGracePeriodSeconds,omitempty"`
+	SecurityContext               *core.PodSecurityContext `json:"securityContext,omitempty"`
+	Priority                      *int32                   `json:"priority,omitempty"`
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/types.go]: PriorityClassName too

File: caas/kubernetes/provider/specs/v2_test.go [hunk 23]
```diff
@@ -0,0 +1,591 @@
...
+var _ = gc.Suite(&v2SpecsSuite{})
+
+var versionHeader = `
+version: 2
+`[1:]
+
+func (s *v2SpecsSuite) TestParse(c *gc.C) {
+
+	specStrBase := versionHeader + `
+omitServiceFrontend: true
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/v2_test.go]: No omitServiceFrontend in v2

File: caas/kubernetes/provider/specs/v2_test.go [hunk 24]
```diff
@@ -0,0 +1,591 @@
...
+    args: ["doIt", "--debug"]
+    workingDir: "/path/to/here"
+    ports:
+      - containerPort: 80
+        name: fred
+        protocol: TCP
+      - containerPort: 443
+        name: mary
+    kubernetes:
+      imagePullPolicy: Always
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/v2_test.go]: Move imagePullPolicy to generic I think

File: caas/kubernetes/provider/specs/v2_test.go [hunk 25]
```diff
@@ -0,0 +1,591 @@
...
+      restricted: 'yes'
+      switch: on
+configMaps:
+  mydata:
+    foo: bar
+    hello: world
+service:
+  annotations:
+    foo: bar
+pod:
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/v2_test.go]: I'm thinking perhaps this "pod" section belongs in the kubernetes.yaml file.
Things like restartPolicy could be generic, but we say securtityContext is k8s specific in the containers section so logically it should be k8s specific here too? I think it's easier to start out with stuff being k8s specific and move across as we determine how to model

File: caas/kubernetes/provider/specs/v2_test.go [hunk 26]
```diff
@@ -0,0 +1,591 @@
...
+    supplementalGroups: [1,2]
+  priority: 30
+  readinessGates:
+    - conditionType: PodScheduled
+  dnsPolicy: ClusterFirstWithHostNet
+serviceAccount:
+  name: build-robot
+  automountServiceAccountToken: true
+  capabilities:
+    roleBinding:
```
Comment (wallyworld) [inline caas/kubernetes/provider/specs/v2_test.go]: We'll need to fix this next PR

**Comment by ycliuhw** [comment]
[podSpecV2](https://trello.com/c/i07Ssvok/156-podspecv2)

**Comment by ycliuhw** [comment]
$$merge$$

**Pull request merged**